fog/ledger/connection/src/router_client.rs

```diff
@@ -0,0 +1,256 @@
+
+        let ch = ChannelBuilder::default_channel_builder(env).connect_to_uri(&uri, &logger);
+        let client = LedgerApiClient::new(ch);
+        let (request_sender, response_receiver) = client
+            .request()
+            .expect("Could not retrieve grpc sender and receiver.");
+
+        Self {
+            logger,
+            attest_cipher: None,
+            _client: client,
+            request_sender,
+            response_receiver,
+            uri,
+            verifier,
+        }
+    }
+
+    fn is_attested(&self) -> bool {
+        self.attest_cipher.is_some()
+    }
+
+    async fn attest(&mut self) -> Result<VerificationReport, Error> {
+        // If we have an existing attestation, nuke it.
+        self.deattest();
+
+        let mut csprng = McRng::default();
+
+        let initiator = Start::new(self.uri.responder_id()?.to_string());
+
+        let init_input = ClientInitiate::<X25519, Aes256Gcm, Sha512>::default();
+        let (initiator, auth_request_output) = initiator.try_next(&mut csprng, init_input)?;
+
+        let attested_message: AuthMessage = auth_request_output.into();
+        let mut request = LedgerRequest::new();
+        request.set_auth(attested_message);
+        self.request_sender
+            .send((request.clone(), grpcio::WriteFlags::default()))
+            .await?;
+
+        let mut response = self
+            .response_receiver
+            .try_next()
+            .await?
+            .ok_or(Error::ResponseNotReceived)?;
+        let auth_response_msg = response.take_auth();
+
+        // Process server response, check if key exchange is successful
+        let auth_response_event =
+            AuthResponseInput::new(auth_response_msg.into(), self.verifier.clone());
+        let (initiator, verification_report) =
+            initiator.try_next(&mut csprng, auth_response_event)?;
+
+        self.attest_cipher = Some(initiator);
+
+        Ok(verification_report)
+    }
+
+    fn deattest(&mut self) {
+        if self.is_attested() {
+            log::trace!(self.logger, "Tearing down existing attested connection.");
+            self.attest_cipher = None;
+        }
+    }
+
+    /// Check one or more key images against the ledger router service
+    pub async fn check_key_images(
+        &mut self,
+        key_images: &[KeyImage],
+    ) -> Result<CheckKeyImagesResponse, Error> {
+        log::trace!(self.logger, "Check key images was called");
+        if !self.is_attested() {
+            let verification_report = self.attest().await;
+            verification_report?;
+        }
+
+        let key_images_queries = key_images
+            .iter()
+            .map(|&key_image| KeyImageQuery {
+                key_image,
+                start_block: 0,
+            })
+            .collect();
+        let key_images_request = CheckKeyImagesRequest {
+            queries: key_images_queries,
+        };
+
+        let aad = vec![];
```

Comment:
Just to confirm- this should be empty? If so, maybe add a comment above explaining why.

Reply:
Yes, this should be empty (there's no AAD associated with the ledger query, at least not currently). I'll add a comment.